[Anthony Tamzil] iP

anthonytamzil opened the pull request:
# Chatter

> Keep chattering - Me

Chatter frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPERFAST_ to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍🏼

And it is **FREE**!

Features:

- [X] Managing Tasks
- [ ] Managing Deadlines (Coming soon)
- [ ] Reminders (Coming soon) 

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

anthonytamzil committed:
Change chatbot name and add greeting and exit methods

anthonytamzil committed:
Echo user commands and exits when user says bye

anthonytamzil committed:
Add ListOfTasks class

anthonytamzil committed:
Add Task class

anthonytamzil committed:
Add method to mark tasks as done

anthonytamzil committed:
Add unmarking task method

anthonytamzil committed:
Add Todo class as a child class of Task

anthonytamzil committed:
Add Deadline class as a subclass of Task

anthonytamzil committed:
Add Event class as a subclass of Task

anthonytamzil committed:
Add Text UI Testing

anthonytamzil committed:
Add exception handling for invalid commands and task adding

anthonytamzil committed:
Add delete method and Java collections to list of tasks

anthonytamzil committed:
Add test cases and fixed bugs

anthonytamzil committed:
Add more test cases for error handling

anthonytamzil committed:
Add file reading feature to chatbot

anthonytamzil committed:
Add task saving to local file feature

anthonytamzil committed:
Merge branch 'branch-Level-7'

anthonytamzil committed:
Add documentation

anthonytamzil committed:
Add date recognition for Deadline events

anthonytamzil committed:
Add localDate storage for events

anthonytamzil committed:
Merge branch 'branch-Level-8'

anthonytamzil committed:
Add commands and OOP structure

anthonytamzil committed:
Add new ExitCommand

anthonytamzil committed:
Add MarkCommand and ListCommand

anthonytamzil committed:
Add DeleteCommand

anthonytamzil committed:
Fix bugs

anthonytamzil committed:
Merge branch 'branch-A-MoreOOP'

anthonytamzil committed:
Divide classes into packages

anthonytamzil committed:
Merge branch 'A-Packages'

anthonytamzil committed:
Merge branch 'add-gradle-support'

anthonytamzil committed:
Set up Gradle

anthonytamzil committed:
Add test cases

anthonytamzil committed:
Add more Storage test cases

anthonytamzil committed:
Merge branch 'branch-A-JUnit'

anthonytamzil committed:
Package app as JAR file

anthonytamzil committed:
Merge branch 'A-Jar'

anthonytamzil committed:
Add FindCommand

anthonytamzil committed:
Merge branch 'branch-Level-9'

anthonytamzil committed:
Add JavaDocs

anthonytamzil committed:
Ensure coding standard

anthonytamzil committed:
Merge branch 'branch-Level-9'

anthonytamzil committed:
Merge branch 'branch-A-CodingStandard'

anthonytamzil committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/chatter/command/FindCommand.java

anthonytamzil committed:
Merge branch 'master' of https://github.com/anthonytamzil/ip

anthonytamzil committed:
Add style changes

JCSnap commented on src/main/java/chatter/Chatter.java:
```diff
@@ -0,0 +1,69 @@
+        try {
+            this.tasks = new TaskList(storage.readFile());
+        } catch (IOException e) {
+            ui.showLoadingError();
```
Good naming to let reader know the details of the function! At first glance, I can already tell that the function call is to display error message for loading, and not just a generic showError

JCSnap commented on src/main/java/chatter/command/AddCommand.java:
```diff
@@ -0,0 +1,41 @@
+     * @param storage chatter.Storage class to read and store tasks by the user.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
```
Not sure if you agree but do you think there is a need to make the function name more descriptive? Eg: addTask instead of just execute

JCSnap commented on src/main/java/chatter/task/Event.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.altStart + convertDateToString(this.start)
```
Very good naming for function! convertDateToString allows me to know the role of the function right away!

JCSnap commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,55 @@
+
+public class ParserTest {
+    @Test
+    public void testParseExit() throws ChatterException {
```
perhaps function names can be improved by referring to the style guide! eg: testFunction_argument_expectedOutput

JCSnap reviewed (commented):
Overall styles follow the convention! I especially like how the functions are named after their purpose and it is very easy for me as a reviewer to follow!

anthonytamzil committed:
Update readme

anthonytamzil committed:
Add javaFX

anthonytamzil committed:
Add chatbot response

anthonytamzil committed:
Add error handling to chatbot

anthonytamzil committed:
Merge branch 'branch-Level-10'

papataco14 commented on src/main/java/chatter/command/Command.java:
```diff
@@ -0,0 +1,51 @@
+import chatter.TaskList;
+import chatter.Ui;
+
+/**
+ * Represents an abstract chatter.command.Command class.
+ *
+ * @author Anthony Tamzil
+ * @version CS2103T Individual Project AY2023/24 Semester 1
+ */
+public abstract class Command {
```
I like your usage of abstract class for the Command class to not allow it to be instantiated

papataco14 commented on src/main/java/chatter/Chatter.java:
```diff
@@ -0,0 +1,87 @@
+            }
+
+            try {
+                Command c = Parser.parse(fullCommand);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
```
Is there a more descriptive name for c?

papataco14 commented on src/main/java/chatter/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+ * @version CS2103T Individual Project AY2023/24 Semester 1
+ */
+public class Deadline extends Task {
+    /** A string indicating deadline of task in localDate form. */
+    protected LocalDate by;
+    /** A string indicating deadline of task not in local date format. */
+    protected String altBy;
```
Are there more descriptive names for by and altBy?

papataco14 commented on src/main/java/chatter/Parser.java:
```diff
@@ -0,0 +1,74 @@
+            return new MarkCommand(false, Character.getNumericValue(fullCommand.charAt(7)));
+        case("delete"):
+            return new DeleteCommand(Character.getNumericValue(fullCommand.charAt(7)));
+        case("todo"):
+            if (fullCommand.length() < 6) {
+                throw new ChatterException("☹ OOPS!!! The description of a todo cannot be empty!");
+            }
+            return new AddCommand(new ToDo(fullCommand.substring(5)));
+        case("deadline"):
+            int deadlineIndex = fullCommand.indexOf("/by");
+            if (deadlineIndex == -1) {
+                throw new ChatterException("Please add a '/by' statement with the deadline.");
+            }
+
+            return new AddCommand(new Deadline(fullCommand.substring(9, deadlineIndex - 1),
+                    fullCommand.substring(deadlineIndex + 4)));
+        case("event"):
+            int startIndex = fullCommand.indexOf("/from");
+            if (startIndex == -1) {
+                throw new ChatterException("Please add a '/from' statement with the start time / date.");
+            }
+
+            int endIndex = fullCommand.indexOf("/to");
+            if (endIndex == -1) {
+                throw new ChatterException("Please add a '/to' statement with the end time / date.");
+            }
+
+            return new AddCommand(new Event(fullCommand.substring(6, startIndex - 1),
+                    fullCommand.substring(startIndex + 6, endIndex - 1),
+                    fullCommand.substring(endIndex + 4)));
+        case("bye"):
+            return new ExitCommand();
+        case("find"):
+            return new FindCommand(fullCommand.substring(5));
+        default:
+            throw new ChatterException("OOPS!!! I'm sorry, but I don't know what that means :-("
+                    + "\nPlease enter a valid command!");
+        }
+    }
+}
```
Good usage of the switch block to handle different cases, but maybe you could use enums to represent the switch cases.

papataco14 reviewed (commented):
LGTM, just some to nits to fix!

anthonytamzil committed:
Fix method names for testcases

Previous method names followed the wrong convention.

Changes the method names to the appropriate convention for test methods.

Following the convention improves code quality and allows other
programmers to easily identify functionality and purpose of the
test method.

anthonytamzil committed:
Add assert statements to Parser class

Assert can be useful to check that assumptions about the program are
true before the program continues.

Let's add assert statements to the Parser class.

Assert statements are important to ensure that the user inputs are as
expected before returning the Command classes.

anthonytamzil committed:
Improve coding standards

anthonytamzil committed:
Merge pull request #2 from anthonytamzil/branch-A-Assertions

Add assertions to code

anthonytamzil committed:
Merge pull request #1 from anthonytamzil/branch-A-CodeQuality

Improve code quality

anthonytamzil committed:
Add Help Command

anthonytamzil committed:
Merge branch 'branch-BCD-Extension'

anthonytamzil committed:
Add README

anthonytamzil committed:
Update build.gradle
